integration-tests/src/tests/client/resharding.rs

```diff
@@ -963,6 +972,62 @@ fn test_shard_layout_upgrade_simple_v2_seed_44() {
     test_shard_layout_upgrade_simple_impl(ReshardingType::V2, 44, false);
 }
 
+fn test_shard_layout_upgrade_gc_impl(resharding_type: ReshardingType, rng_seed: u64) {
+    init_test_logger();
+    tracing::info!(target: "test", "test_shard_layout_upgrade_gc_impl starting");
+
+    let genesis_protocol_version = get_genesis_protocol_version(&resharding_type);
+    let target_protocol_version = get_target_protocol_version(&resharding_type);
+
+    let epoch_length = 5;
+    let mut test_env: TestReshardingEnv = TestReshardingEnv::new(
+        epoch_length,
+        2,
+        2,
+        100,
+        None,
+        genesis_protocol_version,
+        rng_seed,
+        false,
+        Some(resharding_type),
+    );
+
+    // We keep trie data for 5 epochs, so running step about 8 times should be good
+    let drop_chunk_condition = DropChunkCondition::new();
+    for _ in 1..8 * epoch_length {
+        test_env.step(&drop_chunk_condition, target_protocol_version);
+    }
```

Comment:
It would be nice to check that the cleanup happens exactly when you expect it to happen, no sooner no later. Can you test it based on height here?

Reply:
Done!